### ISSUE TITLE
Nextjs14 with Supabase Auth - AuthApiError: invalid claim: missing sub claim

### ISSUE THREAD

**Issue by pdomala** [supabase/supabase-js#992]
# Bug report

- [x] I confirm this is a bug with Supabase, not with my own application. 
- [x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

I am trying to setup Supabase authentication using Google Oauth provider. I am following the instructions details in the below link
https://supabase.com/docs/guides/auth/server-side/nextjs

I created a login page with a button which calls `supabase.auth.signInWithOAuth({ provider: 'google' }`. I see the use being created in Supabase. But when I try to get the session using below code. I get the session as null and the console throws this error `AuthApiError: invalid claim: missing sub claim`

```
const supabase = createClient();
const { data: session, error } = await supabase.auth.getSession();
console.log(session);
console.log(error);
```

```
AuthApiError: invalid claim: missing sub claim
    at handleError (webpack-internal:///(rsc)/./node_modules/@supabase/gotrue-js/dist/module/lib/fetch.js:52:11)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async _handleRequest (webpack-internal:///(rsc)/./node_modules/@supabase/gotrue-js/dist/module/lib/fetch.js:89:9)
    at async _request (webpack-internal:///(rsc)/./node_modules/@supabase/gotrue-js/dist/module/lib/fetch.js:74:18)
    at async eval (webpack-internal:///(rsc)/./node_modules/@supabase/gotrue-js/dist/module/GoTrueClient.js:831:24)
    at async SupabaseAuthClient._useSession (webpack-internal:///(rsc)/./node_modules/@supabase/gotrue-js/dist/module/GoTrueClient.js:754:20)
    at async SupabaseAuthClient._getUser (webpack-internal:///(rsc)/./node_modules/@supabase/gotrue-js/dist/module/GoTrueClient.js:825:20)
    at async eval (webpack-internal:///(rsc)/./node_modules/@supabase/gotrue-js/dist/module/GoTrueClient.js:813:20)
    at async eval (webpack-internal:///(rsc)/./node_modules/@supabase/gotrue-js/dist/module/GoTrueClient.js:699:28) {
  __isAuthError: true,
  status: 401
```

## To Reproduce

- Create a fesh NextJs application
- Follow the instruction at https://supabase.com/docs/guides/auth/server-side/nextjs

## System information

- OS: MacOS
- Browser (if applies) Chrome
- Version of supabase-js: `@supabase/supabase-js": "^2.39.8`
- Next JS Version: `14.1.3`
- Version of Node.js: `v20.10.0`

**Comment by pdomala** [comment]
Adding my sample repo with Supabase where the error can be re-produced.
https://github.com/pdomala/nextjs14-supabase-sample

- Add env with Supabase URL and Anon key
- npm run dev

**Comment by nicolazj** [comment]
I've got the same issue with linkedin provider.  my code worked for a while , then stopped working..

**Comment by nicolazj** [comment]
@pdomala  ok i figured out, you have to have a auth/callback/route to handle the `code` thingy

` await supabase.auth.exchangeCodeForSession(code);`

**Comment by bigdenergy** [comment]
> @pdomala ok i figured out, you have to have a auth/callback/route to handle the `code` thingy
> 
> ` await supabase.auth.exchangeCodeForSession(code);`

Can u send me sample code of your auth/callback/route.ts ? 

Thank you!

**Comment by rookledookle** [comment]
Struggling with this too followed the docs to no avail

**Comment by joris-delorme** [comment]
I got the same issue...

**Comment by ItaiAxelrad** [comment]
Having the same issue. I set up everything according to the docs. The user is signed in successfully upon first sign up but once they log out they are unable to log back in. The error code is 403: invalid claim: missing sub claim.

**Comment by aleskozelsky** [comment]
Same here

**Comment by Daniel-Alamezie** [comment]
Authentication seems to be broken atm. My linkedIn auth provider was broken yesterday, a fix was put out for that this morning. But google Auth provider is broken too.
getting this error
  data: { user: null },
  error: AuthApiError: invalid claim: missing sub claim
      at handleError (webpack-internal:///(rsc)/./node_modules/@supabase/auth-js/dist/module/lib/fetch.js:62:11)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async _handleRequest (webpack-internal:///(rsc)/./node_modules/@supabase/auth-js/dist/module/lib/fetch.js:116:9)
      at async _request (webpack-internal:///(rsc)/./node_modules/@supabase/auth-js/dist/module/lib/fetch.js:92:18)
      at async eval (webpack-internal:///(rsc)/./node_modules/@supabase/auth-js/dist/module/GoTrueClient.js:1064:24)
      at async SupabaseAuthClient._useSession (webpack-internal:///(rsc)/./node_modules/@supabase/auth-js/dist/module/GoTrueClient.js:949:20)
      at async SupabaseAuthClient._getUser (webpack-internal:///(rsc)/./node_modules/@supabase/auth-js/dist/module/GoTrueClient.js:1058:20)
      at async eval (webpack-internal:///(rsc)/./node_modules/@supabase/auth-js/dist/module/GoTrueClient.js:1041:20)
      at async eval (webpack-internal:///(rsc)/./node_modules/@supabase/auth-js/dist/module/GoTrueClient.js:897:28) {
    __isAuthError: true,
    status: 403,
    code: 'bad_jwt'

my callback route looks like this
`
import { cookies } from 'next/headers'
import { NextResponse } from 'next/server'
import { type CookieOptions, createServerClient } from '@supabase/ssr'

export async function GET(request: Request) {
  const { searchParams, origin } = new URL(request.url)
  const code = searchParams.get('code')
  // if "next" is in param, use it as the redirect URL
  const next = searchParams.get('next') ?? '/'
  console.log('code:', code)

  if (code) {
    const cookieStore = cookies()
    const supabase = createServerClient(
      process.env.NEXT_PUBLIC_SUPABASE_URL!,
      process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
      {
        cookies: {
          get(name: string) {
            return cookieStore.get(name)?.value
          },
          set(name: string, value: string, options: CookieOptions) {
            cookieStore.set({ name, value, ...options })
          },
          remove(name: string, options: CookieOptions) {
            cookieStore.delete({ name, ...options })
          },
        },
      }
    )
    const { error } = await supabase.auth.exchangeCodeForSession(code)
    if (!error) {
      return NextResponse.redirect(`${origin}/dashboard`)
    }
    console.log(error);
  }

  // return the user to an error page with instructions
  return NextResponse.redirect(`${origin}/auth/auth-code-error`)
}
`
` async function signInWithGoogle() {
    const supabase = supabaseBrowser();
    const { error } = await supabase.auth.signInWithOAuth({
      provider: 'google',
      options: {
        queryParams: {
          access_type: 'offline',
          prompt: 'consent',
        },
        redirectTo: `${window.location.origin}/auth/callback`,
      }
    });
    if (error) {
      console.error('Error signing in:', error);
    }
  }
`
It sucessfully creates the user, but and logs in but when i try to fetch userId or session i run into the prescribed error

**Comment by Daniel-Alamezie** [comment]
Yeah, its slowing down my development process. The linkedIn one wasnt supabase's fault tbf, LinkedIn had release a breaking change without communicating it and that causes linkedIn auth to be broken. But the fix for that was merged this morning so it might take a while before that gets rolled out. But then my Google auth had also been broken.

**Comment by Daniel-Alamezie** [comment]
Oh, they just env variable no? they should just work. I use them other places in the code and they work fine for their function

**Comment by Daniel-Alamezie** [comment]
Hmm my issue isnt that the authentication and account creation doesnt work, That works fine. The issue i am getting is that, after the code has been extract and an exchange token is set. I cant seem to get a user when I try to query the auth.getUser(). If that makes sense

**Comment by Daniel-Alamezie** [comment]
Like i can see the user data after the account creation in the db, but when i try querying i dont get anything back. And its not a rls issue ive checked

**Comment by Daniel-Alamezie** [comment]
It seems like it was a bug with the api keys. I created a new project to test. My code remained the same and i swapped the api keys with the new supabase project keys and those work with the authentication. Supa weird

**Comment by vasyaqwe** [comment]
Facing the same error, but it only happens if I call supabase.auth.getUser() in an API route handler. If called in a server component, no errors.

**Comment by fabiokandrade** [comment]
> Facing the same error, but it only happens if I call supabase.auth.getUser() in an API route handler. If called in a server component, no errors.

same here!

**Comment by Daniel-Alamezie** [comment]
> Facing the same error, but it only happens if I call supabase.auth.getUser() in an API route handler. If called in a server component, no errors.

Yeah j got around fixing this by making a supavase client that uses supabase url and supabase secret Key. Not anon key, this allows you to use the client in api route

**Comment by mxmzb** [comment]
It actually works fine. I guess you're all mixing up the server and client side supabase clients at some point.

The two resources that you need are these (combine these snippets): 

1. https://supabase.com/docs/guides/auth/server-side/oauth-with-pkce-flow-for-ssr?queryGroups=environment&environment=server
2. https://supabase.com/docs/guides/auth/server-side/nextjs

If you want to make it work on the server side, most everything touching supabase should be server client.

**Comment by mhdevfr** [comment]
I have the same Issue with NUXT 3

**Comment by ERPG** [comment]
This is happening for me as well when calling a request to a route handler from a server component.

**Comment by latifs** [comment]
Any updates on this?

**Comment by mhdevfr** [comment]
> Any updates on this?

nop :(

**Comment by jumploops** [comment]
We ran into the same issue today, but we're still on the older [@supabase/auth-ui-react](https://github.com/supabase-community/auth-ui) library. 
  
The issue appeared when we upgraded React to `18.3.1`. We pinned React to `18.2.0` and everything is back to working as expected.
